google/cloud/spanner_dbapi/transaction_helper.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2023 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import random
+from dataclasses import dataclass
+from enum import Enum
+from typing import TYPE_CHECKING, List, Union, Any
+from google.api_core.exceptions import Aborted
+
+import time
+
+from google.cloud.spanner_dbapi.batch_dml_executor import BatchMode
+from google.cloud.spanner_dbapi.exceptions import RetryAborted
+from google.cloud.spanner_dbapi.parsed_statement import (
+    StatementType,
+    ClientSideStatementType,
+)
+
+if TYPE_CHECKING:
+    from google.cloud.spanner_dbapi import Connection, Cursor
+from google.cloud.spanner_dbapi.checksum import ResultsChecksum, _compare_checksums
+
+MAX_INTERNAL_RETRIES = 50
+RETRY_ABORTED_ERROR = "The transaction was aborted and could not be retried due to a concurrent modification."
+
+
+class TransactionRetryHelper:
+    def __init__(self, connection: "Connection"):
+        """Helper class used in retrying the transaction when aborted This will
+        maintain all the statements executed on original transaction and replay
+        them again in the retried transaction.
+
+        :type connection: :class:`~google.cloud.spanner_dbapi.connection.Connection`
+        :param connection: A DB-API connection to Google Cloud Spanner.
+        """
+
+        self._connection = connection
+        # list of all single and batch statements in the same order as executed
+        # in original transaction along with their checksum value
+        self._statement_result_details_list: List[StatementResultDetails] = []
+        self._last_statement_result_details: StatementResultDetails = None
+
+    def _set_connection_for_retry(self):
+        self._connection._spanner_transaction_started = False
+        self._connection._transaction_begin_marked = False
+        self._connection._batch_mode = BatchMode.NONE
+
+    def reset(self):
+        """
+        Resets the state of the class when the ongoing transaction is committed
+        or aborted
+        """
+        self._statement_result_details_list = []
+        self._last_statement_result_details = None
+
+    def add_fetch_statement_for_retry(self, rows, exception, is_fetch_all):
+        if not self._connection._client_transaction_started:
+            return
+        if (
+            self._last_statement_result_details is not None
+            and self._last_statement_result_details.statement_type
+            == CursorStatementType.FETCH_MANY
```

Comment:
Please see the above comment where we are storing fetchone statement as CursorStatementType.FETCH_MANY. So in case of a new fetch statement we are checking whether the last statement is of type fetchOne or fetchMany and if so then we are just appending the modifying the last statement rather than adding  a new statement in the list to retry